File: OfflineTranscription/ViewModels/TranscriptionViewModel.swift
import Foundation
import SwiftData
import Observation
import AVFoundation

@MainActor
@Observable
final class TranscriptionViewModel {
    let whisperService: WhisperService

    var showError: Bool = false
    var errorMessage: String = ""
    var showPermissionDenied: Bool = false
    private var recordingStartTime: Date?
    private var recordingDuration: TimeInterval = 0

    var isRecording: Bool { whisperService.isRecording }
    var confirmedText: String { whisperService.confirmedText }
    var hypothesisText: String { whisperService.hypothesisText }
    var translatedConfirmedText: String { whisperService.translatedConfirmedText }
    var translatedHypothesisText: String { whisperService.translatedHypothesisText }
    var translationWarning: String? { whisperService.translationWarning }
    var bufferEnergy: [Float] { whisperService.bufferEnergy }
    var bufferSeconds: Double { whisperService.bufferSeconds }
    var tokensPerSecond: Double { whisperService.tokensPerSecond }
    var cpuPercent: Double { whisperService.cpuPercent }
    var memoryMB: Double { whisperService.memoryMB }
    var selectedModel: ModelInfo { whisperService.selectedModel }
    var translationEnabled: Bool { whisperService.translationEnabled }
    var fullText: String { whisperService.fullTranscriptionText }
    var sessionState: SessionState { whisperService.sessionState }

    /// True when the engine has a mid-session error to surface.
    var hasEngineError: Bool { whisperService.lastError != nil }

    /// True when the session was interrupted (phone call, etc.)
    var isInterrupted: Bool { whisperService.sessionState == .interrupted }

    init(whisperService: WhisperService) {
        self.whisperService = whisperService
    }

    func toggleRecording() async {
        if isRecording || isInterrupted {
            stopRecording()
        } else {
            await startRecording()
        }
    }

    func startRecording() async {
        do {
            recordingStartTime = Date()
            recordingDuration = 0
            showPermissionDenied = false
            try await whisperService.startRecording()
        } catch let error as AppError {
            if case .microphonePermissionDenied = error {
                showPermissionDenied = true
            }
            showError = true
            errorMessage = error.localizedDescription
        } catch {
            showError = true
            errorMessage = error.localizedDescription
        }
    }

    func stopRecording() {
        if let start = recordingStartTime {
            recordingDuration = Date().timeIntervalSince(start)
        } else {
            recordingDuration = 0
        }
        whisperService.stopRecording()
    }

    /// Surface any engine error via the shared error alert.
    func surfaceEngineError() {
        if let error = whisperService.lastError {
            if case .translationUnavailable = error {
                whisperService.clearLastError()
                return
            }
            if case .translationFailed = error {
                whisperService.clearLastError()
                return
            }
            showError = true
            errorMessage = error.localizedDescription
            whisperService.clearLastError()
        }
    }

    func openSettings() {
        PermissionManager.openAppSettings()
    }

    func clearTranscription() {
        recordingDuration = 0
        recordingStartTime = nil
        showPermissionDenied = false
        whisperService.clearTranscription()
    }

    #if DEBUG
    func transcribeTestFile(_ path: String) {
        whisperService.transcribeTestFile(path)
    }
    #endif

    @discardableResult
    func saveTranscription(using context: ModelContext) -> Bool {
        let text = fullText
        guard !text.isEmpty else { return false }

        // Use bufferSeconds as fallback for test-file transcriptions where recordingDuration == 0
        let duration = recordingDuration > 0 ? recordingDuration : whisperService.bufferSeconds
        guard duration > 0 else { return false }

        let record = TranscriptionRecord(
            text: text,
            durationSeconds: duration,
            modelUsed: whisperService.selectedModel.displayName
        )

        // Save audio WAV alongside the text record
        let samples = whisperService.currentAudioSamples
        if !samples.isEmpty {
            do {
                try SessionFileManager.saveAudio(samples: samples, for: record.id)
                record.audioFileName = SessionFileManager.relativeAudioPath(for: record.id)
            } catch {
                NSLog("[Save] Failed to write audio WAV: \(error)")
            }
        }

        context.insert(record)
        do {
            try context.save()
            return true
        } catch {
            SessionFileManager.deleteSession(for: record.id)
            showError = true
            errorMessage = "Failed to save: \(error.localizedDescription)"
            return false
        }
    }
}

File: VoicePingIOSAndroidOfflineSpeechTranslationAndroid/app/src/main/kotlin/com/jima/offlinetranscription/ui/transcription/TranscriptionViewModel.kt
package com.voiceping.offlinetranscription.ui.transcription

import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.voiceping.offlinetranscription.data.AppDatabase
import com.voiceping.offlinetranscription.data.TranscriptionEntity
import com.voiceping.offlinetranscription.model.AudioInputMode
import com.voiceping.offlinetranscription.model.ModelInfo
import com.voiceping.offlinetranscription.model.TranslationProvider
import com.voiceping.offlinetranscription.service.WhisperEngine
import com.voiceping.offlinetranscription.util.WavWriter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

class TranscriptionViewModel(
    val engine: WhisperEngine,
    private val database: AppDatabase,
    private val filesDir: File
) : ViewModel() {

    val isRecording = engine.isRecording
    val confirmedText = engine.confirmedText
    val hypothesisText = engine.hypothesisText
    val bufferEnergy = engine.bufferEnergy
    val bufferSeconds = engine.bufferSeconds
    val tokensPerSecond = engine.tokensPerSecond
    val lastError = engine.lastError
    val selectedModel = engine.selectedModel
    val modelState = engine.modelState
    val useVAD = engine.useVAD
    val enableTimestamps = engine.enableTimestamps
    val audioInputMode = engine.audioInputMode
    val systemAudioCaptureReady = engine.systemAudioCaptureReady
    val isSystemAudioCaptureSupported: Boolean
        get() = engine.isSystemAudioCaptureSupported
    val translationEnabled = engine.translationEnabled
    val speakTranslatedAudio = engine.speakTranslatedAudio
    val translationSourceLanguageCode = engine.translationSourceLanguageCode
    val translationTargetLanguageCode = engine.translationTargetLanguageCode
    val ttsRate = engine.ttsRate
    val translatedConfirmedText = engine.translatedConfirmedText
    val translatedHypothesisText = engine.translatedHypothesisText
    val translationWarning = engine.translationWarning
    val translationProvider = engine.translationProvider
    val translationModelReady = engine.translationModelReady
    val translationDownloadStatus = engine.translationDownloadStatus
    val isAndroidSystemTranslationAvailable = engine.isAndroidSystemTranslationAvailable
    val cpuPercent = engine.cpuPercent
    val memoryMB = engine.memoryMB
    val e2eResult = engine.e2eResult

    private val _showSaveConfirmation = MutableStateFlow(false)
    val showSaveConfirmation: StateFlow<Boolean> = _showSaveConfirmation.asStateFlow()

    val fullText: String
        get() = engine.fullTranscriptionText

    private inline fun launchEngineAction(crossinline block: suspend () -> Unit): Job {
        return viewModelScope.launch {
            block()
        }
    }

    private fun copyAssetIfMissing(context: Context, assetName: String): File {
        val cached = File(context.cacheDir, assetName)
        if (cached.exists()) return cached
        context.assets.open(assetName).use { input ->
            cached.outputStream().use { output -> input.copyTo(output) }
        }
        return cached
    }

    private fun sessionDirFor(sessionId: String): File = File(filesDir, "sessions/$sessionId")

    fun toggleRecording() {
        if (engine.isRecording.value) {
            engine.stopRecording()
        } else {
            startRecordingWithPreparation()
        }
    }

    fun startRecordingWithPreparation() {
        launchEngineAction {
            engine.prewarmRealtimePath()
            engine.startRecording()
        }
    }

    fun prewarmOnScreenOpen() {
        launchEngineAction {
            engine.prewarmRealtimePath()
        }
    }

    fun clearTranscription() {
        engine.clearTranscription()
    }

    fun setAudioInputMode(mode: AudioInputMode) {
        engine.setAudioInputMode(mode)
    }

    fun setSystemAudioCapturePermission(resultCode: Int, data: Intent?) {
        engine.setSystemAudioCapturePermission(resultCode, data)
    }

    /** Dismiss error without clearing transcription text. */
    fun dismissError() {
        engine.clearError()
    }

    fun saveTranscription() {
        val text = fullText
        if (text.isBlank()) {
            Log.w("TranscriptionVM", "saveTranscription: fullText is blank, nothing to save")
            return
        }

        // Use actual audio buffer duration, not wall clock
        val duration = engine.recordingDurationSeconds
        val language = engine.detectedLanguage.value
        val modelName = engine.selectedModel.value.displayName
        Log.i("TranscriptionVM", "saveTranscription: text=${text.length} chars, duration=${"%.1f".format(duration)}s, model=$modelName, lang=$language")

        viewModelScope.launch {
            val entity = TranscriptionEntity(
                text = text,
                durationSeconds = duration,
                modelUsed = modelName,
                language = language
            )

            // Write audio WAV if samples are available
            val samples = engine.audioRecorder.samples
            var audioRelPath: String? = null
            if (samples.isNotEmpty()) {
                if (engine.audioRecorder.hasDroppedSamples) {
                    Log.w("TranscriptionVM", "Audio buffer was trimmed during long recording — saved WAV will be incomplete")
                }
                try {
                    val sessionDir = sessionDirFor(entity.id)
                    val wavFile = File(sessionDir, "audio.wav")
                    withContext(Dispatchers.IO) {
                        WavWriter.write(samples, outputFile = wavFile)
                    }
                    audioRelPath = "sessions/${entity.id}/audio.wav"
                    Log.i("TranscriptionVM", "Saved audio WAV: ${wavFile.length()} bytes (${samples.size} samples, ${"%.1f".format(samples.size / 16000.0)}s)")
                } catch (e: Exception) {
                    Log.w("TranscriptionVM", "Failed to write audio WAV", e)
                }
            } else {
                Log.w("TranscriptionVM", "No audio samples available for WAV save")
            }

            try {
                database.transcriptionDao().insert(
                    entity.copy(audioFileName = audioRelPath)
                )
                Log.i("TranscriptionVM", "Session saved to history: id=${entity.id}")
                _showSaveConfirmation.value = true
            } catch (e: Exception) {
                Log.e("TranscriptionVM", "Failed to save session to database", e)
                // Clean up audio file on DB failure
                if (audioRelPath != null) {
                    sessionDirFor(entity.id).deleteRecursively()
                }
                engine.setLastError(
                    com.voiceping.offlinetranscription.model.AppError.TranscriptionFailed(e)
                )
            }
        }
    }

    fun dismissSaveConfirmation() {
        _showSaveConfirmation.value = false
    }

    fun transcribeTestFile(filePath: String) {
        engine.transcribeFile(filePath)
    }

    fun transcribeTestAsset(context: Context) {
        val cached = copyAssetIfMissing(context, "test_speech.wav")
        engine.transcribeFile(cached.absolutePath)
    }

    fun stopIfRecording() {
        if (engine.isRecording.value) {
            engine.stopRecording()
        }
    }

    fun switchModel(model: ModelInfo) {
        launchEngineAction {
            engine.switchModel(model)
        }
    }

    fun setUseVAD(enabled: Boolean) {
        launchEngineAction {
            engine.setUseVAD(enabled)
        }
    }

    fun setEnableTimestamps(enabled: Boolean) {
        launchEngineAction {
            engine.setEnableTimestamps(enabled)
        }
    }

    fun setTranslationEnabled(enabled: Boolean) {
        launchEngineAction {
            engine.setTranslationEnabled(enabled)
        }
    }

    fun setSpeakTranslatedAudio(enabled: Boolean) {
        launchEngineAction {
            engine.setSpeakTranslatedAudio(enabled)
        }
    }

    fun setTranslationSourceLanguageCode(languageCode: String) {
        launchEngineAction {
            engine.setTranslationSourceLanguageCode(languageCode)
        }
    }

    fun setTranslationTargetLanguageCode(languageCode: String) {
        launchEngineAction {
            engine.setTranslationTargetLanguageCode(languageCode)
        }
    }

    fun setTtsRate(rate: Float) {
        launchEngineAction {
            engine.setTtsRate(rate)
        }
    }

    fun setTranslationProvider(provider: TranslationProvider) {
        launchEngineAction {
            engine.setTranslationProvider(provider)
        }
    }
}
